pepper-apis/src/main/java/org/broadinstitute/ddp/service/FileUploadService.java

```diff
@@ -133,51 +137,70 @@ URL generateSignedUrl(String blobName, String mimeType, HttpMethod method) {
                 Storage.SignUrlOption.withExtHeaders(headers));
     }
 
+    // Convenience helper to lock file upload before verifying.
+    public Optional<VerifyResult> verifyUpload(Handle handle, long participantUserId, long uploadId) {
+        return handle.attach(FileUploadDao.class)
+                .findAndLockById(uploadId)
+                .map(upload -> verifyUpload(handle, participantUserId, upload));
+    }
+
     /**
      * Determine if file meets criteria, and update its status appropriately. E.g. file must be uploaded, matches
-     * reported file size, and belong to the participant. Files already marked uploaded will not be checked again.
+     * reported file size, and belong to the participant. Files already verified will not be checked again.
+     *
+     * <p>File upload should be "locked" to ensure its status doesn't change midway or file gets moved to a different
+     * bucket. Since the file upload is locked, caller should commit transaction as soon as possible to unblock other
+     * transactions waiting on the file upload row.
      *
      * @param handle            the database handle
      * @param participantUserId the participant who this upload will be assigned to
      * @param upload            the file upload
      * @return check result
      */
-    public CheckResult checkAndSetUploadStatus(Handle handle, long participantUserId, FileUpload upload) {
+    public VerifyResult verifyUpload(Handle handle, long participantUserId, FileUpload upload) {
```

Comment:
I struggled with implementing the verification here. I think it's mostly due to the asynchronous nature of the file operations. E.g. upload is done asynchronously by client, answering file question happens outside context of upload, files are scanned synchronously somewhere else, file status might change or file gets moved to different bucket, etc.

I think the key here is to lock the file upload row in the database. So when user PATCHes answer, we lock the row and do the verification. In the "file scan result handler" which I'm implementing, it'll try to lock the same file upload before updating scan status or moving the file to a bucket. So hopefully this helps avoid where scenarios.

Reply:
Right. Imagine if the bucket file got moved/deleted right in-between the `select` and `fetchBlob()` call. On the file scan handler side, we'll first try to lock before moving the file.